### PR TITLE
Fix: Some fields in OrtCUDAProviderOptionsV2 struct are not initialized 

### PR DESCRIPTION
**Pull request by snnn**
### Description
The file include/onnxruntime/core/providers/cuda/cuda_provider_options.h is a C++ file. It is not for C. 

Though OrtCUDAProviderOptionsV2 is defined as a struct. It is a normal C++ class. You can use any C++ grammar there.
C++ class normally use constructors to initialize its member variables. 
C normally use memset or specialized functions(e.g. having a function named InitOrtCUDAProviderOptionsV2)

Whatever, you should provide an initialization function at where you define the data structure, so that whatever people add new members to the data structure they would know they also should update the init function.

### Motivation and Context
Some of the members were definitely not initialized in some cases.

### CODE REVIEW

**Comment by yuslepukhin** [comment]
The struct is no longer compliable by a C compiler after this change, which I think was the original intent. It is likely used in the C API.
Without initialization, it represents a POD and the constructor is not generated even when compiled by a CPP compiler, with the initialization it is no longer a POD

**Review by yuslepukhin — APPROVED** [review]

File: include/onnxruntime/core/providers/cuda/cuda_provider_options.h [hunk 1]
```diff
@@ -14,21 +14,21 @@
...
-  int enable_skip_layer_norm_strict_mode;                  // flag specifying if SkipLayerNorm is in strict mode. If true, use LayerNormalization kernel.
-                                                           // The strict mode has better accuracy but lower performance.
+  int device_id = 0;                                                                                           // cuda device id.
+  int has_user_compute_stream = 0;                                                                             // indicator of user specified CUDA compute stream.
+  void* user_compute_stream = nullptr;                                                                         // user specified CUDA compute stream.
+  int do_copy_in_default_stream = 1;                                                                           // flag specifying if the default stream is to be used for copying.
+  OrtCudnnConvAlgoSearch cudnn_conv_algo_search = OrtCudnnConvAlgoSearchExhaustive;                            // cudnn algo search enum.
+  size_t gpu_mem_limit = std::numeric_limits<size_t>::max();                                                   // BFC Arena memory limit for CUDA.
```
Comment (yuslepukhin) [inline include/onnxruntime/core/providers/cuda/cuda_provider_options.h:22]: I am not sure gpu limit is the same as x64 host memory limit, unless it is just a senintel token

Comment (snnn) [reply include/onnxruntime/core/providers/cuda/cuda_provider_options.h:22]: I just copied the code from the old place. If it is a concern, we can address it separately.

**Review by yuslepukhin — APPROVED** [review]

File: include/onnxruntime/core/providers/cuda/cuda_provider_options.h (hunk 1 above)
Comment (snnn) [reply include/onnxruntime/core/providers/cuda/cuda_provider_options.h:22]: Here it means: "Do not set a limit on the max size of the arena." @souptc, is this the best way to express "no limit"?

**Comment by snnn** [comment]
Before my change, this header file is already not compatible with C compilers. Because it has:
```
onnxruntime::ArenaExtendStrategy arena_extend_strategy;
```
The "onnxruntime::" string is a namespace. C doesn't have namespaces. 

And this file is intended to be internal only. It is an internal header file. It should not be included in onnxruntime_c_api.h and should not be used by when using the public C APIs. User can only get the instance of OrtCUDAProviderOptionsV2 via CreateCUDAProviderOptions. In such a way we can add new members to this struct without breaking binary compatibility. 
Since it is an internal header,  we can safely use C++ grammar there.

**Comment by pranavsharma** [comment]
This file is not meant to be used directly by our users. People ought to be using the C API or the corresponding C++ wrapper. For API users, this is an opaque struct that is created/updated/deleted using CreateCUDAProviderOptions, UpdateCUDAProviderOptions, ReleaseCUDAProviderOptions.

**Pull request merged**